feat: use Connected App for OAuth based Email Account

revant opened the pull request:
TODO

- [x] Manual tests with Google
- [x] Manual tests with Microsoft
- [x] Code review / improvement
- [x] Automated tests?
- [x] Patches

I have just replaced the oauth token logic from `OAuth` class (`email/oauth.py`) with `Connected App`, `Token Cache`

revant committed:
feat: use Connected App for OAuth based Email Account

revant commented on frappe/email/doctype/email_account/email_account.js:
```diff
@@ -68,15 +68,19 @@ frappe.email_defaults_pop = {
 
 function oauth_access(frm) {
```
this nested callback needs to improve.

Alchez replied:
If we don't want to make another whitelisted method outside of the controller class, I think you can use `with_doc` and `get_doc` to do it.

```javascript
function oauth_access(frm) {
	frappe.model.with_doc("Connected App", frm.doc.connected_app, () => {
		const connected_app = frappe.get_doc("Connected App", frm.doc.connected_app);
		return frappe.call({
			doc: connected_app,
			method: "initiate_web_application_flow",
			callback: function(r) {
				window.open(r.message, "_self");
			}
		});
	})
}
```

revant commented on frappe/email/doctype/email_account/email_account.js:
```diff
@@ -167,21 +169,25 @@ frappe.ui.form.on("Email Account", {
-	after_save(frm) {
-		if (frm.doc.auth_method === "OAuth" && !frm.doc.refresh_token) {
-			oauth_access(frm);
+		if (frm.doc.connected_app && !frm.doc.connected_user) {
```
any better way to set session user in user link field?

Alchez replied:
Related question: would an account need to be authorized by administrator for background pulls to succeed?

revant replied:
I am assuming same functionality as current Email Account where System Manager adds an `Email Account` or `Email Domain` with needed addresses and ports and sets awaiting password so the user can just fill up their password and continue.

In OAuth, System Manager maps `Connected App` and Frappe `User` in `Email Account`. User clicks authorize on Email Account and logins in with account on Connected App (Google/Microsoft/Examplemail).

Do we need Admin to further do anything?

Alchez commented on frappe/email/doctype/email_account/email_account.js:
```diff
@@ -197,13 +203,22 @@ frappe.ui.form.on("Email Account", {
-			frm.dashboard.set_headline_alert(msg, "yellow");
-		}
+		frappe.call({
+			method: "frappe.integrations.doctype.connected_app.connected_app.check_active_token",
```
This should only happen if the auth method is OAuth. I'm seeing errors on basic-auth accounts as well.

revant committed:
fix(email): oauth_access call and info message

lxnow commented:
Let's keep this alive please! Frappe customer waiting for fix.

revant commented:
someone try it out. I'm not in enough pain to go through pain of client registration on google and microsoft and get keys.

If anyone provides me ready keys or tries with their keys there will be progress.

git-avc commented:
I can provide you keys for M365. 
reva**.**e@g***l.com ?

revant commented:
for quick interaction and discussion use telegram group to find me.

ahmadpak commented:
Any update @git-avc @revant

git-avc commented:
Testing it. Work in progress.

revant commented:
Faced this for microsoft. https://stackoverflow.com/a/46948056.

Got refresh_token by setting env variable `OAUTHLIB_RELAX_TOKEN_SCOPE=True`. Should we just hardcode it in pycode?

For containers, passing variable is not a problem. For native setup I think it needs to be done in supervisor config.

EDIT: Authorization Code Flow and Token Refresh is working. Email Account code is messed up. I'm doing `frappe.db.set_value` to change email account values for testing. SSL/TLS server fields, port fields somehow became disabled.

lxnow commented:
bump

ankush commented:
It will happen when it happens. Dont follow up on PRs. :pray:

phot0n committed:
fix: make connected app work with email account

* removed (now) unnecessary things from oauth class
* simplified get_active_token in connected_app
* removed gmail banner from email account docs

phot0n commented:
working with gmail. not sure about microsoft.

https://user-images.githubusercontent.com/32034600/209805224-fcdad6d1-20e0-4746-be87-ab0052c8006a.mp4

phot0n commented on frappe/email/doctype/email_account/email_account.js:
```diff
@@ -170,39 +157,19 @@ frappe.ui.form.on("Email Account", {
-		if (frm.doc.service === "GMail") {
-			frm.dashboard.set_headline_alert(msg);
-		}
-	},
```
I've removed this banner - let me know if this needs to be there.

revant replied:
Banner can be removed.

phot0n committed:
refactor(minor): simplify check_active_token

phot0n committed:
fix: redirect back to email account after successful oauth

phot0n committed:
chore: handle get requests via frappe.whitelist decorator

phot0n commented on frappe/email/doctype/email_account/email_account.js:
```diff
@@ -170,39 +160,19 @@ frappe.ui.form.on("Email Account", {
-			);
-			frm.dashboard.clear_headline();
-			frm.dashboard.set_headline_alert(msg, "yellow");
+		if (frm.doc.auth_method === "OAuth") {
+			frappe.call({
+				method: "frappe.integrations.doctype.connected_app.connected_app.check_active_token",
+				args: {
+					connected_app: frm.doc.connected_app,
+					connected_user: frm.doc.connected_user,
+				},
+				callback: (r) => {
+					if (!r.message) {
+						oauth_access(frm);
+					}
+				},
+			});
```
this is not *really* required - we anyway do the same when testing for connection upon saving an email account and the new flow involves setting up the connected app before proceeding to the email account setup part

also the `check_active_token` function doesn't really check if the token is active or not - it basically refreshes the token if it's expired

revant replied:
we can remove the js part.

`check_active_token` can be renamed to `fetch_active_token`?

phot0n committed:
chore: remove tracking changes for token cache

happens very frequently for email and leads to unnecessary bloat in version doctype
plus token caches are only read and delete only

phot0n committed:
fix: get_expires_in logic for token cache

phot0n committed:
chore: log any exception while refreshing access token in connected apps

phot0n commented on frappe/integrations/doctype/connected_app/connected_app.py:
```diff
@@ -102,17 +102,39 @@ def get_scopes(self):
+		if token_cache and token_cache.is_expired():
+			oauth_session = self.get_oauth2_session(user)
+
+			try:
+				token = oauth_session.refresh_token(
+					body=f"redirect_uri={self.redirect_uri}",
+					token_url=self.token_uri,
+					refresh_token=token_cache.get_password("refresh_token"),
+				)
+			except Exception as e:
+				msg = _("An error occurred while refreshing the access token for Connected App {0}: ").format(
+					self.name
+				)
+				title = "Oauth Token Refresh Error"
+				frappe.throw(msg + str(e), title=_(title))
+				self.log_error(title, str(e))
```
currently only added logging in the case any exception occurs while refreshing the access token.

i'm kinda skeptical about the expiry of refresh token as the user will have to again authorize to get a new refresh token - any thoughts on this are welcome.

ref (for google): https://developers.google.com/identity/protocols/oauth2#expiration - they've not stated anything for prod usage so i'm kinda guessing the token will remain active for a long time (unless the user does someting)

revant replied:
silent logging is fine. emails are also sent in background from workers

Whatever the refresh limits are the will be followed. If there is no refresh limit (generally in case of production, verified by provider clients) the emails work forever, unless user goes to google's connected apps section and removes/revokes the client.

phot0n reviewed (commented):
migration will be a problem - as people will have to add/change their whitelisted callbacks on google console.

rest all can hopefully be done - we can pick the client id and secret from `Google Settings` doctype and the access and refresh tokens from the email accounts itself and put an expiry of 0 so that they immediately get refreshed on the next use via token cache

phot0n committed:
fix: only validate oauth if tokens are set

* also brough back oauth authorization message

phot0n commented:
@git-avc I've unlocked conversation on this thread.

> I'm trying to test https://github.com/frappe/frappe/pull/18750 ... (sorry, comments there are closed xD ), with Microsoft365 mail account IMAP + Oauth2 I use this repo https://github.com/revant/frappe/tree/fix-oauth2-and-connected-app-tests

> It's the right one?

Nope - the branch would be `oauth-email` (https://github.com/revant/frappe/tree/oauth-email)

phot0n commented on frappe/integrations/doctype/token_cache/token_cache.json:
```diff
@@ -106,5 +107,5 @@
  ],
  "sort_field": "modified",
  "sort_order": "DESC",
- "track_changes": 1
```
changes happen very frequently in token cache documents in case of emails - hence removed this

git-avc commented:
I've tried with M365 unsuccesfully . Can't retrieve email with OAuth ... it works from Thunderbird +OAuth.
I get this on error log:

```
Traceback with variables (most recent call last):
  File "apps/frappe/frappe/email/doctype/email_account/email_account.py", line 498, in get_inbound_mails
    if email_server.select_imap_folder(folder.folder_name):
      process_mail = <function EmailAccount.get_inbound_mails.<locals>.process_mail at 0x7fa4ac04c670>
      email_sync_rule = 'UID 1:251'
      email_server = <frappe.email.receive.EmailServer object at 0x7fa4aab9e4a0>
      folder = <IMAPFolder: 414eb9f354 parent=revant>
      mails = []
      self = <EmailAccount: revant>
  File "apps/frappe/frappe/email/receive.py", line 169, in select_imap_folder
    res = self.imap.select(f'"{folder}"')
      self = <frappe.email.receive.EmailServer object at 0x7fa4aab9e4a0>
      folder = 'INBOX'
  File "/usr/lib/python3.10/imaplib.py", line 756, in select
    typ, dat = self._simple_command(name, mailbox)
      self = <frappe.email.receive.Timed_IMAP4_SSL object at 0x7fa4aa187e80>
      mailbox = '"INBOX"'
      readonly = False
      name = 'SELECT'
  File "/usr/lib/python3.10/imaplib.py", line 1230, in _simple_command
    return self._command_complete(name, self._command(name, *args))
      self = <frappe.email.receive.Timed_IMAP4_SSL object at 0x7fa4aa187e80>
      name = 'SELECT'
      args = ('"INBOX"',)
  File "/usr/lib/python3.10/imaplib.py", line 1055, in _command_complete
    raise self.error('%s command error: %s %s' % (name, typ, data))
      self = <frappe.email.receive.Timed_IMAP4_SSL object at 0x7fa4aa187e80>
      name = 'SELECT'
      tag = b'JHKK2'
      logout = False
      typ = 'BAD'
      data = [b'User is authenticated but not connected.']
imaplib.error: SELECT command error: BAD [b'User is authenticated but not connected.']
```

On M365 side, auth is ok on logs ...

tcpaddock commented:
> I've tried with M365 unsuccesfully . 

I was able to successfully connect SMTP and IMAP with M365. There were 2 actions that seemed to be the trick:

1. Separating the email API permissions into its own app registration. Only select Graph permissions 'IMAP.AccessAsUser.All' and 'SMTP.Send'.
2. In Frappe, ensure to use scopes 
https://outlook.office.com/IMAP.AccessAsUser.All and 
https://outlook.office.com/SMTP.Send according to the [documentation](https://learn.microsoft.com/en-us/exchange/client-developer/legacy-protocols/how-to-authenticate-an-imap-pop-smtp-application-by-using-oauth#get-an-access-token).

There seems to be some sort of hybrid scoping happening where office.com scopes are required but adding other Microsoft Graph api permissions prevent the office.com prefix.

Token refresh is not working for me yet.

tcpaddock commented:
For M365 token refresh, it seems you need to add 'offline_access' scope to the request but the response does not return it, only the refresh token. This feature is giving me server error and message:
`Warning: Scope has changed from "https://outlook.office.com/IMAP.AccessAsUser.All offline_access https://outlook.office.com/SMTP.Send" to "https://outlook.office.com/IMAP.AccessAsUser.All https://outlook.office.com/SMTP.Send".`

@phot0n perhaps ignoring the scope change or add checkbox field to ignore scope in response?

git-avc commented:
> For M365 token refresh, it seems you need to add 'offline_access' scope to the request but the response does not return it, only the refresh token.

We've faced this too ...

git-avc commented:
> > I've tried with M365 unsuccesfully .
> 
> I was able to successfully connect SMTP and IMAP with M365. There were 2 actions that seemed to be the trick:
> 
> 1. Separating the email API permissions into its own app registration. Only select Graph permissions 'IMAP.AccessAsUser.All' and 'SMTP.Send'.
> 2. In Frappe, ensure to use scopes
>    https://outlook.office.com/IMAP.AccessAsUser.All and
>    https://outlook.office.com/SMTP.Send according to the [documentation](https://learn.microsoft.com/en-us/exchange/client-developer/legacy-protocols/how-to-authenticate-an-imap-pop-smtp-application-by-using-oauth#get-an-access-token).

I have this permissions too in app registration:
![image](https://user-images.githubusercontent.com/94137451/210797032-329cb569-2075-4ee5-a702-4ba9233e639d.png)

But doesn't works for me ... keep trying.

revant commented:
> Got refresh_token by setting env variable `OAUTHLIB_RELAX_TOKEN_SCOPE=True`. Should we just hardcode it in pycode?

This environment variable along with offline_access is needed to get refresh token.

phot0n committed:
fix: only pull from email accoutns which have token

* removed unnecessary after_save client hook from email account
* renamed check_active_token -> has_token (no need to refresh the access token)
* removed oauthentication error from email oauth - now just a simple validation error

phot0n commented:
Things seem to be working properly for microsoft as well - a minor hiccup: microsoft doesnt like if we have a different email address (sender) from the one email account is authenticated with (or it could be just that `example.com` has been blacklisted by them :P)

we can just check this option to send the mails (present in email account):
![Screenshot from 2023-01-13 17-04-45](https://user-images.githubusercontent.com/32034600/212311452-8d8e21b3-2640-439f-a6fa-75487e13e672.png)

OR we can also setup user emails in user document(s) - this way the user can choose from which email they want to send the email with/from.

also, we need the env variable which @revant mentioned in his earlier [comment](https://github.com/frappe/frappe/pull/18750#issuecomment-1372461771) - we can just hardcode this imo; dont like the idea of adding another column (checkbox) to set/unset this (?)

phot0n committed:
fix: use OAUTHLIB_RELAX_TOKEN_SCOPE for ignoring scope change

without this we get an error regarding the mismatch of scopes from microsoft

git-avc commented:
> Things seem to be working properly for microsoft as well - a minor hiccup: microsoft doesnt like if we have a different email address (sender) from the one email account is authenticated with (or it could be just that `example.com` has been blacklisted by them :P)
> 
> we can just check this option to send the mails (present in email account): ![Screenshot from 2023-01-13 17-04-45](https://user-images.githubusercontent.com/32034600/212311452-8d8e21b3-2640-439f-a6fa-75487e13e672.png)
> 
> OR we can also setup user emails in user document(s) - this way the user can choose from which email they want to send the email with/from.
> 
> also, we need the env variable which @revant mentioned in his earlier [comment](https://github.com/frappe/frappe/pull/18750#issuecomment-1372461771) - we can just hardcode this imo; dont like the idea of adding another column (checkbox) to set/unset this (?)

It seems working now on Microsoft 365! Sending mails from one mailbox using other mail address requires a permission on Exchange, called "Send on behalf"

https://learn.microsoft.com/en-us/microsoft-365/admin/add-users/give-mailbox-permissions-to-another-user?view=o365-worldwide

Thanks to @revant @phot0n @ankush @Alchez @tcpaddock and everyone else involved, for your efforts to make it working!

phot0n committed:
chore: minor cleanup - removed unnecessary branches and comments

phot0n committed:
chore(patch): disable all email accounts with oauth mechanism

phot0n commented:
I've added a patch to disable all email accounts with oauth enabled so that people can resetup them with connected app flow - even if we did copy over the creds from google settings and try to migrate stuff, there is human intervention required (for whitelisting, etc) hence, thought it's better to just let people restup them.

lemme know if that seems fine or not :)

phot0n commented:
Thanks everyone for the efforts!

barredterra commented:
Is a v13 backport feasible?
